AX: Web content can become inaccessible when revealed by an animating dropdown component

twilco opened the pull request:
#### 0af03dfdd5d39c45999aba3e13cd86a2f2f69051
<pre>
AX: Web content can become inaccessible when revealed by an animating dropdown component
<a href="https://bugs.webkit.org/show_bug.cgi?id=277160">https://bugs.webkit.org/show_bug.cgi?id=277160</a>
<a href="https://rdar.apple.com/132590239">rdar://132590239</a>

Reviewed by Chris Fleizach.

This bug happened due to this sequence:

  1. A button is pressed, animating the dropdown to become open (changing display:none to display:block and quickly
     increasing its height from zero to full-height in rapid &quot;ticks&quot;).

  2. The container of the dropdown and button dynamically becomes ignored, as it no longer has only inline children
     (an esoteric condition in `AccessibilityRenderObjet::computeAccessibilityIsIgnored`:
     <a href="https://github.com/WebKit/WebKit/blob/e27ca2b56be9822117da0ea79062e38e8598cd23/Source/WebCore/accessibility/AccessibilityRenderObject.cpp#L1307)">https://github.com/WebKit/WebKit/blob/e27ca2b56be9822117da0ea79062e38e8598cd23/Source/WebCore/accessibility/AccessibilityRenderObject.cpp#L1307)</a>

  3. This means the effective parent of the button and dropdown has changed (they have a new unignored parent).

  4. In AXIsolatedTree::updateChildren, we prevent deletion of reparented objects via m_protectedFromDeletionIDs.

  5. However, the bug is that we clear this list any time we `AXIsolatedTree::queueRemovalsLocked`, rather than once
     every tree update cycle (which is essentially by `AXIsolatedTree::queueAppendsAndRemovals`).

  6. Because we fail to protect the button from deletion, it is removed from `m_nodeMap` during
     `AXIsolatedTree::removeSubtreeFromNodeMap` in the middle of `AXIsolatedTree::updateChildren`.

  7. In `AXIsolatedTree::collectNodeChangesForSubtree`, we properly detect that the effective parent for the button
     has changed, and queue an entry for it in `AXIsolatedTree::m_needsParentUpdate`.

  8. When we process that list in the final `AXIsolatedTree::queueAppendsAndRemovals`, we queue an entry to
     `m_pendingParentUpdates` that sets the new parent of the button to the invalid `AXID()`, because we try to get the
     new parent of the object from the nodemap, but it was deleted from the nodemap in step 6.

  9. The object becomes completely detached from the accessibility tree (having a parent ID of 0), breaking VoiceOver
     navigation almost entirely.

Fix this by only clearing `m_protectedFromDeletionIDs` once every tree update cycle (i.e. in
`AXIsolatedTree::queueAppendsAndRemovals`).

This scenario can be triggered by jQuery&apos;s slideToggle method (as discovered on a real webpage).

* LayoutTests/accessibility/animated-dropdown-expected.txt: Added.
* LayoutTests/accessibility/animated-dropdown.html: Added.
* LayoutTests/accessibility/resources/jquery-3.6.1.js: Added.
* LayoutTests/platform/glib/TestExpectations: Disable new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/animated-dropdown-expected.txt: Added.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::queueAppendsAndRemovals):

Canonical link: <a href="https://commits.webkit.org/281440@main">https://commits.webkit.org/281440@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f32c57dc4183bb5c81dca5890f28a2bbcfe50dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59864 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63781 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10388 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61993 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10563 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48541 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7265 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36571 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51832 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29385 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33277 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9069 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9311 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55203 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9349 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65511 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9218 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55884 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3803 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51811 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56027 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13282 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3150 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35023 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36107 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37192 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fleizach reviewed (approved):
Nice find. Thanks for getting a test for this one

twilco committed:
AX: Web content can become inaccessible when revealed by an animating dropdown component
https://bugs.webkit.org/show_bug.cgi?id=277160
rdar://132590239

Reviewed by Chris Fleizach.

This bug happened due to this sequence:

  1. A button is pressed, animating the dropdown to become open (changing display:none to display:block and quickly
     increasing its height from zero to full-height in rapid "ticks").

  2. The container of the dropdown and button dynamically becomes ignored, as it no longer has only inline children
     (an esoteric condition in `AccessibilityRenderObjet::computeAccessibilityIsIgnored`:
     https://github.com/WebKit/WebKit/blob/e27ca2b56be9822117da0ea79062e38e8598cd23/Source/WebCore/accessibility/AccessibilityRenderObject.cpp#L1307)

  3. This means the effective parent of the button and dropdown has changed (they have a new unignored parent).

  4. In AXIsolatedTree::updateChildren, we prevent deletion of reparented objects via m_protectedFromDeletionIDs.

  5. However, the bug is that we clear this list any time we `AXIsolatedTree::queueRemovalsLocked`, rather than once
     every tree update cycle (which is essentially by `AXIsolatedTree::queueAppendsAndRemovals`).

  6. Because we fail to protect the button from deletion, it is removed from `m_nodeMap` during
     `AXIsolatedTree::removeSubtreeFromNodeMap` in the middle of `AXIsolatedTree::updateChildren`.

  7. In `AXIsolatedTree::collectNodeChangesForSubtree`, we properly detect that the effective parent for the button
     has changed, and queue an entry for it in `AXIsolatedTree::m_needsParentUpdate`.

  8. When we process that list in the final `AXIsolatedTree::queueAppendsAndRemovals`, we queue an entry to
     `m_pendingParentUpdates` that sets the new parent of the button to the invalid `AXID()`, because we try to get the
     new parent of the object from the nodemap, but it was deleted from the nodemap in step 6.

  9. The object becomes completely detached from the accessibility tree (having a parent ID of 0), breaking VoiceOver
     navigation almost entirely.

Fix this by only clearing `m_protectedFromDeletionIDs` once every tree update cycle (i.e. in
`AXIsolatedTree::queueAppendsAndRemovals`).

This scenario can be triggered by jQuery's slideToggle method (as discovered on a real webpage).

* LayoutTests/accessibility/animated-dropdown-expected.txt: Added.
* LayoutTests/accessibility/animated-dropdown.html: Added.
* LayoutTests/accessibility/resources/jquery-3.6.1.js: Added.
* LayoutTests/platform/glib/TestExpectations: Disable new test.
* LayoutTests/platform/ios/TestExpectations: Enable new test.
* LayoutTests/platform/ios/accessibility/animated-dropdown-expected.txt: Added.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::queueRemovalsLocked):
(WebCore::AXIsolatedTree::queueAppendsAndRemovals):

Canonical link: https://commits.webkit.org/281440@main
